Include/internal/pycore_stackref.h

```diff
@@ -8,187 +8,214 @@ extern "C" {
-    return cleared;
-}
+#define Py_TAG_PTR      (0)
+#define Py_TAG_BITS     (1)
+
+#ifdef Py_GIL_DISABLED
+    static const _PyStackRef Py_STACKREF_NULL = { .bits = 0 | Py_TAG_DEFERRED};
 #else
-#   define PyStackRef_Get(tagged) ((PyObject *)((tagged).bits))
+    static const _PyStackRef Py_STACKREF_NULL = { .bits = 0 };
 #endif
 
-// Converts a PyObject * to a PyStackRef, stealing the reference.
-#if defined(Py_GIL_DISABLED)
-static inline _PyStackRef
-_PyStackRef_StealRef(PyObject *obj)
-{
-    // Make sure we don't take an already tagged value.
-    assert(((uintptr_t)obj & Py_TAG_DEFERRED) == 0);
-    return ((_PyStackRef){.bits = ((uintptr_t)(obj))});
-}
-#   define PyStackRef_StealRef(obj) _PyStackRef_StealRef(_PyObject_CAST(obj))
+#define PyStackRef_IsNull(stackref) ((stackref).bits == Py_STACKREF_NULL.bits)
+
+
+#ifdef Py_GIL_DISABLED
+#   define PyStackRef_True ((_PyStackRef){.bits = ((uintptr_t)&_Py_TrueStruct) | Py_TAG_DEFERRED })
 #else
-#   define PyStackRef_StealRef(obj) ((_PyStackRef){.bits = ((uintptr_t)(obj))})
+#   define PyStackRef_True ((_PyStackRef){.bits = ((uintptr_t)&_Py_TrueStruct) })
 #endif
 
-// Converts a PyObject * to a PyStackRef, with a new reference
-#if defined(Py_GIL_DISABLED)
-static inline _PyStackRef
-_PyStackRef_NewRefDeferred(PyObject *obj)
+#ifdef Py_GIL_DISABLED
+#   define PyStackRef_False ((_PyStackRef){.bits = ((uintptr_t)&_Py_FalseStruct) | Py_TAG_DEFERRED })
+#else
+#   define PyStackRef_False ((_PyStackRef){.bits = ((uintptr_t)&_Py_FalseStruct) })
+#endif
+
+#ifdef Py_GIL_DISABLED
+#   define PyStackRef_None ((_PyStackRef){.bits = ((uintptr_t)&_Py_NoneStruct) | Py_TAG_DEFERRED })
+#else
+#   define PyStackRef_None ((_PyStackRef){.bits = ((uintptr_t)&_Py_NoneStruct) })
+#endif
+
+
+static inline int
+PyStackRef_Is(_PyStackRef a, _PyStackRef b) {
+    return a.bits == b.bits;
+}
+
+static inline int
+PyStackRef_IsDeferred(_PyStackRef ref)
 {
-    // Make sure we don't take an already tagged value.
-    assert(((uintptr_t)obj & Py_TAG_DEFERRED) == 0);
-    assert(obj != NULL);
-    if (_PyObject_HasDeferredRefcount(obj)) {
-        return (_PyStackRef){ .bits = (uintptr_t)obj | Py_TAG_DEFERRED };
-    }
-    else {
-        return (_PyStackRef){ .bits = (uintptr_t)Py_NewRef(obj) };
-    }
+    return ((ref.bits & Py_TAG_BITS) == Py_TAG_DEFERRED);
+}
+
+#ifdef Py_GIL_DISABLED
+// Gets a PyObject * from a _PyStackRef
+static inline PyObject *
+PyStackRef_AsPyObjectBorrow(_PyStackRef tagged)
+{
+    PyObject *cleared = ((PyObject *)((tagged).bits & (~Py_TAG_BITS)));
+    return cleared;
 }
-#   define PyStackRef_NewRefDeferred(obj) _PyStackRef_NewRefDeferred(_PyObject_CAST(obj))
 #else
-#   define PyStackRef_NewRefDeferred(obj) PyStackRef_NewRef(((_PyStackRef){.bits = ((uintptr_t)(obj))}))
+// Need to define as macro because WASI has very sensitive stack sizes.
+#   define PyStackRef_AsPyObjectBorrow(tagged) ((PyObject *)(tagged).bits)
```

Comment:
Could you fix the warning here?

```none
initialization of ‘PyObject **’ {aka ‘struct _object **’} from incompatible pointer type ‘PyObject *’ {aka ‘struct _object *’} [-Wincompatible-pointer-types]
```

Reply:
Even better- now that WASI is run with -O3, I can just make it a static inline function.

Reply:
You can close this.